A88-xds-data-error-handling.md

```diff
@@ -0,0 +1,538 @@
+  error if it already has a valid cached resource.
+
+- `OnResourceDoesNotExist()`: Invoked specifically for the does-not-exist
+  case.  The watcher is generally expected to stop using any previously
+  cached resource and put itself into a failing state.
+  - Note: The mechanism described in [gRFC A53] inhibits these
+    notifications for only the case where we already had a cached version
+    of the resource, which in the SotW protocol applies only to LDS
+    and CDS.  That mechanism does not affect the case where we do not
+    already have a cached version of the resource, which is where the
+    does-not-exist timer is used.
+
+These methods do not map well to the desired error handling behavior
+defined above.
+
+This proposal replaces those methods with the following two methods:
+
+- `OnResourceChanged(StatusOr<ResourceType> resource)`: Will be invoked
+  to notify the watcher of a new version of the resource received from
+  the xDS server or an error indicating the reason why the resource cannot
+  be obtained.  Note that if an error is passed to this method after the
+  watcher has previously seen a valid resource, the watcher is expected
+  to stop using that previously delivered resource.  In this case, the
+  XdsClient will remove the resource from its cache, so that CSDS (see
+  [gRFC A40]) and XdsClient metrics (see [gRFC A78]) will not reflect a
+  resource that the client is not actually using.
+
+- `OnAmbientError(Status status)`: Will be invoked to notify the watcher
+  of an error that occurs after a resource has been received that should
+  not modify the watcher's use of that resource but that may be useful
+  information about the ambient state of the XdsClient.  In particular,
+  the watcher should *not* stop using the previously seen resource, and the
+  XdsClient will *not* remove the resource from its cache.  However, the
+  error message may be useful as additional context to include in errors
+  that are being generated for other reasons.  For example, if failing a
+  data plane RPC due to all endpoints being unreachable, it may be useful
+  to also report that the client has lost contact with the xDS server.
+  The error condition is considered to be cleared when either
+  `OnResourceChanged()` is invoked or `OnAmbientError()` is invoked
+  again with an OK status.
+
+### Handling Data Errors
+
+As described above, the behavior that we want on data errors depends
+on whether we can depend on the control plane to properly alert human
+operators when these errors occur.  To configure this, we will introduce a
+new server feature in the bootstrap config called "fail_on_data_errors",
+which indicates that we cannot depend on the control plane to properly
+alert human operators.
+
+This server feature affects all data errors that can occur when a
+resource is already cached, which are the following cases:
+- NACKs
+- resource deletions in LDS and CDS
+- [xRFC TP3] errors with status NOT_FOUND or PERMISSION_DENIED
```

Comment:
This (the paragraph below) is the main area that I was hoping to include a reference to the "Propagating status" section.  `NOT_FOUND` is mentioned here explicitly as a status that could result in us deleting the resource, which should pretty much have the potential to fail RPCs.  So a special call-out that the `NOT_FOUND` status should _not_ be propagated directly, and to see the below section for proper handling of the errors seems useful.  Maybe at the very end of that paragraph after it says to call the `OnResourceChanged` method?

Reply:
Done.